### DISCUSSION TITLE
What does create a 'keys' config exactly mean?

### DISCUSSION THREAD

**Discussion by elrhaff** [urql-graphql/urql#1263] [Q&A] [answered]
Hi,

I'm still graphql newbie, so this questions might be a bit stupid. I have a query that looks like this: 

```
query Posts($limit: Int!, $cursor: String){
    posts(limit: $limit, cursor: $cursor){
        hasMore
        posts {
            id
            createdAt
            updatedAt
            title
            descriptionSnippet
            url
            creatorId
        }
    }
}
```
And I get this warning on the frontend:
```
urql-exchange-graphcache.mjs?8108:54 Invalid key: The GraphQL query at the field at `Query.posts({"limit":10})` has a selection set, but no key could be generated for the data at this field.
You have to request `id` or `_id` fields for all selection sets or create a custom `keys` config for `PaginatedPosts`.
Entities without keys will be embedded directly on the parent entity. If this is intentional, create a `keys` config for `PaginatedPosts` that always returns null.
(Caused At: "Posts" query)
```

That this mean that I should add a "keys" field that always returns null from the backend resolver, and add this field to the query?

Thanks in advance for the help.

**Comment by elrhaff** [comment, marked as answer]
Figured it out, I needed to add
```
keys: {
    PaginatedPosts: () => null,
  }
```
to the `cacheExchange` opts constructor.

**Reply by JoviDeCroock** [reply]
Exactly, the explanation can be found [here](https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/)

**Reply by dohomi** [reply]
@JoviDeCroock in Hasura I regularly create n:m relations with tables which consists of two fields which are a combined key. What is the recommended way if two fields combined are the unique key? 
```js
keys: {
  table_mn: (data) => `${data.key1_id};${data.key2_id}`
}
```
Is this the correct approach? I would need to make sure to publish both mn related data keys, is that how urql would always be able to index the mn relation correctly?
Thanks for your insight
